snap-build/xbuild.sh

```diff
@@ -0,0 +1,96 @@
+#!/bin/bash
+#
+# Copyright (c) 2018 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+# Build in parallel snap images using VMs.
+# This script runs in the host.
+
+source lib.sh
+
+export WORKDIR=$(pwd)
+seed_dir=${WORKDIR}/seed
+seed_img=${WORKDIR}/seed.img
+id_rsa_file=${WORKDIR}/id_rsa
+id_rsa_pub_file=${WORKDIR}/id_rsa.pub
+snap_sh=${WORKDIR}/snap.sh
+ssh="ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -i ${id_rsa_file}"
+scp="scp -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o IdentitiesOnly=yes -i ${id_rsa_file}"
+
+gen_seed() {
+	rm -f "${seed_img}"
+	truncate --size 2M "${seed_img}"
+	mkfs.vfat -n cidata "${seed_img}" &> /dev/null
+
+	if [ -n "${http_proxy}" ]; then
+		apt_proxy="apt:\n  https_proxy: http://proxy-chain.intel.com:911\n  proxy: http://proxy-chain.intel.com:911"
+		docker_env="\"HTTP_PROXY=${http_proxy}\" \"HTTPS_PROXY=${https_proxy}\" \"NO_PROXY=${no_proxy}\""
+		env="no_proxy=${no_proxy}\n\
+    NO_PROXY=${no_proxy}\n\
+    http_proxy=${http_proxy}\n\
+    HTTP_PROXY=${http_proxy}\n\
+    https_proxy=${https_proxy}\n\
+    HTTPS_PROXY=${https_proxy}"
+	fi
+
+	docker_dns="$(get_dns)"
+
+	[ ! -f "${id_rsa_file}" ] && ssh-keygen -t rsa -f ${id_rsa_file} -P '' &> /dev/null
+	ssh_key="$(cat ${id_rsa_pub_file})"
+
+	sed \
+		-e "s|@USER@|""${USER}""|g" \
+		-e "s|@SSH_KEY@|""${ssh_key}""|g" \
+		-e "s|@APT_PROXY@|""${apt_proxy}""|g" \
+		-e "s|@DOCKER_ENV@|""${docker_env}""|g" \
+		-e "s|@DOCKER_DNS@|""${docker_dns}""|g" \
+		-e "s|@ENV@|""${env}""|g" \
+		${seed_dir}/user-data.in > ${seed_dir}/user-data
+
+	mcopy -oi "${seed_img}" ${seed_dir}/user-data  ${seed_dir}/meta-data ::
+}
+
+poweroff_and_die() {
+	ip="$1"
+	port="$2"
+	${ssh} "${ip}" -p "${port}" sudo poweroff
+	die "$3"
+}
+
+build_arch() {
+	arch="$1"
+	ip="$(get_ip)"
+	port="$(get_port)"
+	${WORKDIR}/${arch}/run_qemu.sh "${ip}" "${port}" "${id_rsa_file}"
+	${scp} -P "${port}" "${snap_sh}" "${ip}:~/"
+	if [ $? != 0 ]; then
+		poweroff_and_die "${ip}" "${port}" "Could not copy snap script"
+	fi
+
+	${ssh} "${ip}" -p "${port}" "~/snap.sh"
+	if [ $? != 0 ]; then
+		poweroff_and_die "${ip}" "${port}" "Failed to run build script"
+	fi
+
+	${scp} -P "${port}" "${ip}:~/packaging/*.snap" .
+	if [ $? != 0 ]; then
+		poweroff_and_die "${ip}" "${port}" "Failed to get snap image"
+	fi
+
+	${ssh} "${ip}" -p "${port}" sudo poweroff
+}
+
+main() {
+	gen_seed
+	(build_arch amd64 &> ${WORKDIR}/amd64/log) &
```

Comment:
It would be useful to support a `-h` option and potentially also to be able to cross-build for a single named architecture rather than all of them.